File: 05_pgm_abstraccion/src/co/edu/campusucc/poo/Vehiculo.java
package co.edu.campusucc.poo;

public abstract class Vehiculo {

    public Vehiculo() {}

    //Properties
    private String marca;
    private String tipo;
    private String carroseria;

    public String toString() {
        return "Abstract Class --> Vehiculo";
    }

    //Abstract Behavior 
    public abstract void startingEngine();
    public abstract void stopingEngine();

    //Defined Behavior
    public void speedUp(){
        String speedUpString = "⏩ speed up in Progress: ";
        for (int i = 0; i < 20 ; i++) {
            speedUpString +="⏩";
            System.out.println(speedUpString);
            try {
                Thread.sleep(1000-(i*50));
            } catch (InterruptedException e) {
                System.err.println("🛑:"+ e);
            }
        }
    }

    public void parking() throws InterruptedException{
        for (int i = 0; i < 10 ; i++) {
            System.out.println("¡¡🚥 Parking in Progress 🚥!!");
            Thread.sleep(1000-(i*50));
        }
    }

    public void breaking() throws InterruptedException{
        for (int i = 0; i < 10 ; i++) {
            System.out.println("¡¡🚥 Braking in Progress 🚥!!");
            Thread.sleep(1000-(i*100));
        }
    }

    //Getters and Setters
    public String getMarca() {
        return marca;
    }

    public void setMarca(String marca) {
        this.marca = marca;
    }

    public String getTipo() {
        return tipo;
    }

    public void setTipo(String tipo) {
        this.tipo = tipo;
    }

    public String getCarroseria() {
        return carroseria;
    }

    public void setCarroseria(String carroseria) {
        this.carroseria = carroseria;
    }
}

File: c++/05-pgm-abstracion/Vehiculo.cpp
#include <iostream>
#include <string>
#include <thread>
#include <chrono>

class Vehiculo {
protected:
    std::string marca;
    std::string carroseria;
    std::string tipo;
    int velocidad;
public:
    Vehiculo() {
        velocidad = 0;
    }

    void setMarca(std::string m) {
        marca = m;
    }

    void setCarroseria(std::string c) {
        carroseria = c;
    }

    void setTipo(std::string t) {
        tipo = t;
    }

    std::string toString() {
        return "Marca: " + marca + ", Carroseria: " + carroseria + ", Tipo: " + tipo;
    }

    void startingEngine() {
        std::cout << "Starting Engine...\n";
        for (int i = 0; i < 3; i++) {
            std::cout << "Engine Starting in " << 3-i << "...\n";
            std::this_thread::sleep_for(std::chrono::milliseconds(1000));
        }
        std::cout << "Engine Started!\n";
    }

    void stopingEngine() {
        std::cout << "Stopping Engine...\n";
        for (int i = 0; i < 3; i++) {
            std::cout << "Engine Stopping in " << 3-i << "...\n";
            std::this_thread::sleep_for(std::chrono::milliseconds(1000));
        }
        std::cout << "Engine Stopped!\n";
    }

    void speedUp() {
        std::cout << "Speeding Up...\n";
        for (int i = 0; i < 10; i++) {
            velocidad += 10;
            std::cout << "Current Speed: " << velocidad << "\n";
            std::this_thread::sleep_for(std::chrono::milliseconds(500));
        }
    }

    void breaking() {
        std::cout << "Breaking...\n";
        for (int i = 0; i < 5; i++) {
            velocidad -= 10;
            if (velocidad < 0) {
                velocidad = 0;
            }
            std::cout << "Current Speed: " << velocidad << "\n";
            std::this_thread::sleep_for(std::chrono::milliseconds(500));
        }
    }

    void currentSpeed() {
        std::cout << "Current Speed: " << velocidad << "\n";
    }

    void parking() {
        std::cout << "Parking...\n";
    }
};
